### ISSUE TITLE
Fetching the whole device tree is not possible - fault "endless_cycle - The provision seems to be repeating indefinitely" occurs

### ISSUE THREAD

**Issue by okraits** [zaidka/genieacs#247]
I am not able to fetch the whole device tree without getting the fault "endless_cycle - The provision seems to be repeating indefinitely" after some time.

Here's the provision I'm using:

```
// do a firmware update
...
// configure the device
...
// refresh the whole device tree
declare("Tags.refreshed", null, {value: true});

declare("InternetGatewayDevice.*", {value: Date.now()});
declare("InternetGatewayDevice.*.*", {value: Date.now()});
declare("InternetGatewayDevice.*.*.*", {value: Date.now()});
declare("InternetGatewayDevice.*.*.*.*", {value: Date.now()});
declare("InternetGatewayDevice.*.*.*.*.*", {value: Date.now()});
declare("InternetGatewayDevice.*.*.*.*.*.*", {value: Date.now()});
declare("InternetGatewayDevice.*.*.*.*.*.*.*", {value: Date.now()});
```

This provision is run from a preset only if tags != refreshed.

I looked at the different parameters in https://github.com/zaidka/genieacs/blob/master/lib/session.coffee#L636 when the fault appears.
The parameter causing the fault is _sessionContext.iteration_ with a value of 128. It occurs when fetching the 3rd level of the device tree. The value of _sessionContext.iteration_ gets higher from level to level, though.

I then gradually increased the limit, peaking at
```
sessionContext.iteration >= MAX_ITERATIONS * (sessionContext.cycle + 9)
```
The provision then didn't produce a fault anymore. The device tree in the device detail page seems to be complete.

* With v1.0 fetching the whole device tree was no problem, with v1.1 it's not possible at the moment.
* How come the limit for iteration is _MAX_ITERATIONS * (sessionContext.cycle + 1)_? (with _MAX_ITERATIONS_ being 64)
* How can we solve this issue? Is it safe to increase the limit like I did? Or may there be any side effects?

**Comment by akcoder** [comment]
You can refresh the entire tree without issue by using refreshObject
instead of using a declare. You would have to do this via the API instead
of via provisioning scripts though.

**Comment by okraits** [comment]
In the long run I want to do this with the API from a CPE management application. But now I want to use genieacs alone with provisions. But refreshing from provisions not working makes the whole new feature of provisions pretty useless.

**Comment by okraits** [comment]
I just tried to do only the firmware update with the provision and then do the refresh via a preset with a "Refresh InternetGatewayDevice" configuration. The preset with the refresh configuration is run after the preset with the firmware update provision.

Nevertheless the fault still occurs because the provision for the firmware update still seems to run in the background although all statements in the provision have been processed.

Is there a way to finish/exit a provision?

**Comment by okraits** [comment]
Just for the record, the endless cycle fault also appears when creating a refreshObject task via the REST API.

**Comment by vinhthang** [comment]
I also got this, today I will spend my time on session.coffee file for result

`2017-09-12T03:41:45.939Z [INFO] undefined; id="15e742c5665174" acsRequest={"name":"GetParameterNames","parameterPath":"","nextLevel":true} cwmpVersion="1.0"
2017-09-12T03:41:46.435Z [INFO] 192.168.1.164 18d071-H646FW-DSNW6a290900: my ACS request; acsRequestId="15e742c5665175" acsRequestName="GetParameterNames"
2017-09-12T03:41:46.435Z [INFO] undefined; id="15e742c5665175" acsRequest={"name":"GetParameterNames","parameterPath":"","nextLevel":true} cwmpVersion="1.0"
2017-09-12T03:41:46.955Z [INFO] 192.168.1.164 18d071-H646FW-DSNW6a290900: my ACS request; acsRequestId="15e742c5665176" acsRequestName="GetParameterNames"
2017-09-12T03:41:46.956Z [INFO] undefined; id="15e742c5665176" acsRequest={"name":"GetParameterNames","parameterPath":"","nextLevel":true} cwmpVersion="1.0"
2017-09-12T03:41:47.416Z [INFO] 192.168.1.164 18d071-H646FW-DSNW6a290900: my ACS request; acsRequestId="15e742c5665177" acsRequestName="GetParameterNames"
2017-09-12T03:41:47.416Z [INFO] undefined; id="15e742c5665177" acsRequest={"name":"GetParameterNames","parameterPath":"","nextLevel":true} cwmpVersion="1.0"
2017-09-12T03:41:47.915Z [INFO] 192.168.1.164 18d071-H646FW-DSNW6a290900: my ACS request; acsRequestId="15e742c5665178" acsRequestName="GetParameterNames"
2017-09-12T03:41:47.915Z [INFO] undefined; id="15e742c5665178" acsRequest={"name":"GetParameterNames","parameterPath":"","nextLevel":true} cwmpVersion="1.0"
2017-09-12T03:41:48.423Z [INFO] 192.168.1.164 18d071-H646FW-DSNW6a290900: my ACS request; acsRequestId="15e742c5665179" acsRequestName="GetParameterNames"
2017-09-12T03:41:48.423Z [INFO] undefined; id="15e742c5665179" acsRequest={"name":"GetParameterNames","parameterPath":"","nextLevel":true} cwmpVersion="1.0"
2017-09-12T03:41:48.972Z [INFO] 192.168.1.164 18d071-H646FW-DSNW6a290900: my ACS request; acsRequestId="15e742c566517a" acsRequestName="GetParameterNames"
2017-09-12T03:41:48.973Z [INFO] undefined; id="15e742c566517a" acsRequest={"name":"GetParameterNames","parameterPath":"","nextLevel":true} cwmpVersion="1.0"
2017-09-12T03:41:49.510Z [INFO] 192.168.1.164 18d071-H646FW-DSNW6a290900: my ACS request; acsRequestId="15e742c566517b" acsRequestName="GetParameterNames"
2017-09-12T03:41:49.510Z [INFO] undefined; id="15e742c566517b" acsRequest={"name":"GetParameterNames","parameterPath":"","nextLevel":true} cwmpVersion="1.0"
2017-09-12T03:41:50.039Z [INFO] 192.168.1.164 18d071-H646FW-DSNW6a290900: my ACS request; acsRequestId="15e742c566517c" acsRequestName="GetParameterNames"
2017-09-12T03:41:50.040Z [INFO] undefined; id="15e742c566517c" acsRequest={"name":"GetParameterNames","parameterPath":"","nextLevel":true} cwmpVersion="1.0"
2017-09-12T03:41:50.542Z [INFO] 192.168.1.164 18d071-H646FW-DSNW6a290900: my ACS request; acsRequestId="15e742c566517d" acsRequestName="GetParameterNames"
2017-09-12T03:41:50.542Z [INFO] undefined; id="15e742c566517d" acsRequest={"name":"GetParameterNames","parameterPath":"","nextLevel":true} cwmpVersion="1.0"
2017-09-12T03:41:51.061Z [INFO] 192.168.1.164 18d071-H646FW-DSNW6a290900: my ACS request; acsRequestId="15e742c566517e" acsRequestName="GetParameterNames"
2017-09-12T03:41:51.061Z [INFO] undefined; id="15e742c566517e" acsRequest={"name":"GetParameterNames","parameterPath":"","nextLevel":true} cwmpVersion="1.0"

2017-09-12T03:41:51.561Z [ERROR] 192.168.1.164 18d071-H646FW-DSNW6a290900: Channel has faulted; channel="task_59b7506acdad951f698cd3c2" retries=2 faultCode="session endless_cycle" faultMessage="The provision seems to be repeating indefinitely"

`

**Comment by okraits** [comment]
I just tested the refreshObject issue with the latest changes in the master branch (ed873cdd4a5c4d3949e24acfcdc2ad5f32554c4e) and now the channel faults with the message

    Too many commit iterations

How is one supposed to react to this error message?

For the record, I got this by creating a refreshObject task for the root node (InternetGatewayDevice) in the GenieACS GUI device details page.

**Comment by zaidka** [comment]
@okraits can you share your device's data model so I can try reproducing this using the simulator? You can use the dump tool under tools directory or use mongodump.

**Comment by okraits** [comment]
@zaidka Sent you a mail with the data model. Thanks in advance for trying to reproduce the issue!

**Comment by mths0x5f** [comment]
I have found this issue too. Interested in helping.

**Comment by okraits** [comment]
I haven't got any real solution either. At the moment I work around the issue by raising the limits in session.coffee until no further fault appears.

**Comment by rfdrake** [comment]
It looks like this was added in 2e40ef86 which is to prevent loops in preset logic.

My theory is that raising the limit only would cause problems if loops are common.  I raised MAX_ITERATIONS to 256 and it seems to have fixed my problems.

**Comment by zaidka** [comment]
@rfdrake so you still need to modify MAX_ITERATIONS even after said commit?

**Comment by kollyma** [comment]
We experience the same issue. GenieACS was installed using NPM: 
```
#npm ls | head  -n 1
genieacs@1.1.1 /usr/lib/node_modules/genieacs
```
Where can we set the MAX_ITERATIONS? Do we need to compile it from source to set the variable?  I found the setting in session.js: 
```
#sed -n 43p /usr/lib/node_modules/genieacs/lib/session.js
MAX_ITERATIONS = 64;
```

**Comment by rfdrake** [comment]
Yes, I raised it to prevent early faults.  I didn't see them very often, but if I asked to refresh a big tree like Device.* it would fault.
